sway-core/src/type_system/info.rs

```diff
@@ -1156,6 +1156,43 @@ impl TypeInfo {
         matches!(self, TypeInfo::Tuple(_))
     }
 
+    pub fn contains_never(&self, engines: &Engines) -> bool {
+        match self {
+            TypeInfo::Never => true,
+            TypeInfo::Tuple(type_arguments) => type_arguments
+                .iter()
+                .any(|t| engines.te().get(t.type_id).contains_never(engines)),
+            TypeInfo::Array(type_argument, _) => engines
+                .te()
+                .get(type_argument.type_id)
+                .contains_never(engines),
+            TypeInfo::Unknown
+            | TypeInfo::UnknownGeneric { .. }
+            | TypeInfo::Placeholder(_)
+            | TypeInfo::TypeParam(_)
+            | TypeInfo::StringSlice
+            | TypeInfo::StringArray(_)
+            | TypeInfo::UnsignedInteger(_)
+            | TypeInfo::Enum(_)
+            | TypeInfo::Struct(_)
+            | TypeInfo::Boolean
+            | TypeInfo::ContractCaller { .. }
+            | TypeInfo::Custom { .. }
+            | TypeInfo::B256
+            | TypeInfo::Numeric
+            | TypeInfo::Contract
+            | TypeInfo::ErrorRecovery(_)
+            | TypeInfo::Storage { .. }
+            | TypeInfo::RawUntypedPtr
+            | TypeInfo::RawUntypedSlice
+            | TypeInfo::Ptr(_)
```

Comment:
Also here, `Ptr`, `Slice`, `Alias`, and `Ref`. If the types they point to or contain or alias contain `Never` then I think they also cannot be inhabited.